### ISSUE TITLE
Call Recording

### ISSUE THREAD

**Issue by atharen** [vector-im/element-call#283]
_**Using MoSCoW method:**_
_M - Must have; S - Should have; C - Could have; W - Won't have_

**Is your feature request related to a problem? Please describe.**
I would like a convenient and reliable way to record calls.

**Describe the solution you'd like**
[M] At minimum, the ability to click a button during a call session, initiating a local recording, including the video and audio of all participants, and if screen sharing is applicable, the audio playing through that respective stream.

Other, lower priority features to consider (Probably can be split in their own tasks/issues...): 
- [S] Choice of what to record (Browser tab, specific screen, specific window)
- [S] Placing notifications that a call has been initiated to existing users, and any new users who join later on (Sort of consent to being recorded).
- [S] Specify location of where to save output
- [C] Video output parameters would be nice to have (bitrate, resolution, etc..)
- [C] Expand on storage options by allowing to save to cloud services (Nextcloud, Google Drive, etc..)
- [C] Should chat be added, record a log of that as well.

**Describe alternatives you've considered**
OBS is an option, but not really convenient.
[Jitsi](https://meet.jit.si/) has call recordings but only through Dropbox.
Zoom has call recordings but closed source.

**Additional context**
[Recording through WebRTC](https://www.webrtc-experiment.com/RecordRTC/) could be an option...
I think this would cover the first lower priority bullet as should be built-in natively by the browser.

Could be made available as an option in the ellipses menu.
![record_button](https://user-images.githubusercontent.com/23165475/160497567-db64f777-ed1c-41b1-b7d6-24a9b0c87b04.png)

_Recording_ settings could go in an appropriate new tab
![record_settings](https://user-images.githubusercontent.com/23165475/160497743-aa663adc-34af-403a-9ee1-010d2a4a1cca.png)

**Comment by compgeniuses** [comment]
The call recording could also offer an option to store recording on Nextcloud, or download the recording after the call.

**Comment by toniocat** [comment]
How about the party who's call you are recording could also set a flag "Do not allow my call to be recorded" as well?  This could work well for copyright video or audio, that shouldn't be recorded.

**Comment by mingsterism** [comment]
is there any updates on this feature? how can it be done if i'd wish to contribute

**Comment by fkwp** [comment]
> is there any updates on this feature? how can it be done if i'd wish to contribute

There are several options to tackle that issue:
- Distributed local recording
  - Every participant does store a local copy of its webcam/audio/screen-sharing streams including metadata (at least timestamps). After the meeting those streams are encrypted & uploaded to the media repository (or next cloud or whatever)
  - Requires a kind of video-conferencing player application or post-processing to replay the video-conference
- Recording bot
  - Note due to MatrixRTC being E2EE this recording bot needs to be part (a "regular" participant) of the conference in order to receive the media E2EE keys.
  - This bot would record all media streams and then render a final video.

**Comment by mingsterism** [comment]
which would you recommend is the better option for the long term. seems to me like the Recording bot will be best as it does not store data on people's own local storage which can take space. 
if it's a recording bot, do you have an idea of how it could be done? 
sorry, i'm new to element-call. happy to look deeper into the codes but if there was some direction that would be greatly appreciated as well. 

thanks

**Comment by fkwp** [comment]
In very secure aware environments I would recommend the local recording.

For the implementation of the recording bot there are on a high level two different approaches:
- Using a headless web browser and capture screen including audio.
- Writing a backend bot component (a MatrixRTC client) 
   - record all media streams and 
   - then render a final video

**Comment by iaxat** [comment]
Is there any update on this?

**Comment by gmanskibiditoilet** [comment]
its great feature that should have been added in the 1st place!

**Comment by Sulfiore** [comment]
LiveKit supports recordings with it's egress plugin, it uploads to s3 and other methods
https://docs.livekit.io/transport/media/ingress-egress/egress/
This would make it a lot easier

LiveKit Meet does already do this, so there is even an example

**Comment by fkwp** [comment]
> LiveKit supports recordings with it's egress plugin, it uploads to s3 and other methods https://docs.livekit.io/transport/media/ingress-egress/egress/ This would make it a lot easier
> 
> LiveKit Meet does already do this, so there is even an example

we are using E2EE of media streams -- this won't work

**Comment by nevtonnnn** [comment]
I've just implemented such a mautrix bot that uses egress. It can only record calls in rooms without E2EE, but I think this is much better than nothing.
https://github.com/nevtonnnn/matrixRTC-recording-bot/tree/master
Please give me feedback and criticism on this bot so I can improve it.
I've also noticed that Element client developers have added an experimental flag [feature_disable_call_per_sender_encryption](https://github.com/element-hq/element-web/blob/develop/docs/labs.md#:~:text=Disable%20per%2Dsender%20encryption%20for%20Element%20Call%20(feature_disable_call_per_sender_encryption)). In my experience, this flag still doesn't work, but I believe that in the near future it will be possible to make calls without E2EE in E2EE rooms.

**Comment by fkwp** [comment]
> I've just implemented such a mautrix bot that uses egress. It can only record calls in rooms without E2EE, but I think this is much better than nothing. https://github.com/nevtonnnn/matrixRTC-recording-bot/tree/master Please give me feedback and criticism on this bot so I can improve it. I've also noticed that Element client developers have added an experimental flag [feature_disable_call_per_sender_encryption](https://github.com/element-hq/element-web/blob/develop/docs/labs.md#:~:text=Disable%20per%2Dsender%20encryption%20for%20Element%20Call%20(feature_disable_call_per_sender_encryption)). In my experience, this flag still doesn't work, but I believe that in the near future it will be possible to make calls without E2EE in E2EE rooms.

Very nice project. However, be aware that we will very soon enable Multi-SFU mode following the MatrixRTC MSCs in Element Call. This will at least break the egress based recording architecture for the embedded / widget use-case if participants of the call are federated.

**Comment by nevtonnnn** [comment]
> Very nice project. However, be aware that we will very soon enable Multi-SFU mode following the MatrixRTC MSCs in Element Call. This will at least break the egress based recording architecture for the embedded / widget use-case if participants of the call are federated.

Thanks for the feedback! But it sounds like the only truly reliable way to implement a production-ready bot is either to fork and rewrite egress as part of the Matrix bot, or to write a custom call recording backend

**Comment by fkwp** [comment]
> > Very nice project. However, be aware that we will very soon enable Multi-SFU mode following the MatrixRTC MSCs in Element Call. This will at least break the egress based recording architecture for the embedded / widget use-case if participants of the call are federated.
> 
> Thanks for the feedback! But it sounds like the only truly reliable way to implement a production-ready bot is either to fork and rewrite egress as part of the Matrix bot, or to write a custom call recording backend

Yes, "sadly" that's the case. I do not recommend to fork egress as the maintenance burden might be substantially. So yes, this boils down writing a bot with one of the following stacks:
- NodeJS, livekit node, matrix-js-sdk, Element Call SDK target
- Matrix rust-sdk (+adding a MatrixRTC layer to it), livekit rust-sdk

For a rust based one https://codeberg.org/esoteric_programmer/matrix-jukebox might be a starting point. However, I think the MatrixRTC implementation is (for good reasons) also lacking Multi-SFU as well as the Matrix 2.0 bits and pieces.

I know, it's a bit of an unfortunate situation. I am pretty confident it gets better once the Matrix 2.0 spec is in FCP and we do have full certainty wrt. MatrixRTC.
